docs/src/tutorials/nonlinear/operator_ad.jl

```diff
@@ -0,0 +1,314 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,       #src
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE    #src
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         #src
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  #src
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  #src
+# SOFTWARE.                                                                      #src
+
+# # Automatic differentiation of user-defined operators
+
+# The purpose of this tutorial is to demonstrate how to apply automatic
+# differentiation to [User-defined operators](@ref jump_user_defined_operators).
+
+# !!! tip
+#     This tutorial is for advanced users. As an alternative, consider using
+#     [Function tracing](@ref) instead of creating an operator, and if an
+#     operator is necessary, consider using the default of `@operator(model, op_f, N, f)`
+#     instead of passing explicit [Gradients and Hessians](@ref).
+
+# This tutorial uses the following packages:
+
+using JuMP
+import Enzyme
+import ForwardDiff
+import Ipopt
+import Test
+
+# As a simple example, we consider the Rosenbrock function:
+
+f(x::T...) where {T} = (1 - x[1])^2 + 100 * (x[2] - x[1]^2)^2
+
+# Here's the value at a random point:
+
+x = rand(2)
+
+#-
+
+f(x...)
+
+# ## Analytic derivative
+
+# If expressions are simple enough, you can provide analytic functions for the
+# gradient and Hessian.
+
+# ### Gradient
+
+# The Rosenbrock function has the gradient vector:
+
+function analytic_∇f(g::AbstractVector, x...)
+    g[1] = 400 * x[1]^3 - 400 * x[1] * x[2] + 2 * x[1] - 2
+    g[2] = 200 * (x[2] - x[1]^2)
+    return
+end
+
+# Let's evaluate it at the same vector `x`:
+
+analytic_g = zeros(2)
+analytic_∇f(analytic_g, x...)
+analytic_g
+
+# ### Hessian
+
+# The Hessian matrix is:
+
+function analytic_∇²f(H::AbstractMatrix, x...)
+    H[1, 1] = 1200 * x[1]^2 - 400 * x[2] + 2
+    ## H[1, 2] = -400 * x[1] <-- not needed because Hessian is symmetric
+    H[2, 1] = -400 * x[1]
+    H[2, 2] = 200.0
+    return
+end
+
+# Note that because the Hessian is symmetric, JuMP requires that we fill in only
+# the lower triangle.
+
+analytic_H = zeros(2, 2)
+analytic_∇²f(analytic_H, x...)
+analytic_H
+
+# ### JuMP example
+
+# Putting our analytic functions together, we get:
+
+function analytic_rosenbrock()
+    model = Model(Ipopt.Optimizer)
+    set_silent(model)
+    @variable(model, x[1:2])
+    @operator(model, op_rosenbrock, 2, f, analytic_∇f, analytic_∇²f)
+    @objective(model, Min, op_rosenbrock(x[1], x[2]))
+    optimize!(model)
+    Test.@test is_solved_and_feasible(model)
+    return value.(x)
+end
+
+analytic_rosenbrock()
+
+# ## ForwardDiff
+
+# Instead of analytic functions, you can use [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl)
+# to compute derivatives.
```

Comment:
Isn't that done by default?
https://jump.dev/JuMP.jl/previews/PR3713/manual/nonlinear/#Automatic-differentiation-2

Reply:
Yes. I guess we should mention that. I added the section more for clarity and as an example.

Reply:
Done